File: services/tradelog/tradelog-api/src/test/java/com/example/tradelog/api/validator/ValidateOptionJournalModelTest.java
//package com.example.tradelog.api.validator;
//
//import com.example.tradelog.api.core.model.ActionType;
//import com.example.tradelog.api.core.model.OptionJournalModel;
//import com.example.tradelog.api.core.model.OptionType;
//import com.example.tradelog.api.core.model.TransactionModel;
//import com.example.tradelog.api.core.model.TransactionSettingsModel;
//import com.example.tradelog.api.core.model.TransactionType;
//import org.junit.jupiter.api.Assertions;
//import org.junit.jupiter.api.Test;
//
//import java.time.OffsetDateTime;
//
//class ValidateOptionJournalModelTest {
//
//    @Test
//    void testValidateOptionJournalModelWhenNull() {
//        Assertions.assertFalse(RequestValidation.validateShareJournalModel.test(null));
//    }
//
//    @Test
//    void testValidateOptionJournalModelWhenActionIsInvalid() {
//        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
//        TransactionModel transactionModel = new TransactionModel("1234", "", OffsetDateTime.now(), "XYZ", TransactionType.OPTION, 6.1f, transactionSettingsModel);
//        OptionJournalModel model = new OptionJournalModel(transactionModel, 1.1f, 2.1f, OffsetDateTime.now(), 1, 5.1f, ActionType.UNKNOWN, OptionType.PUT);
//
//        Assertions.assertFalse(RequestValidation.validateOptionJournalModel.test(model));
//    }
//
//    @Test
//    void testValidateOptionJournalModelWhenActionTypeIsUnknown() {
//        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
//        TransactionModel transactionModel = new TransactionModel("1234", "", OffsetDateTime.now(), "XYZ", TransactionType.OPTION, 6.1f, transactionSettingsModel);
//        OptionJournalModel model = new OptionJournalModel(transactionModel, 1.1f, 2.1f, OffsetDateTime.now(), 1, 5.1f, ActionType.BUY, OptionType.UNKNOWN);
//        Assertions.assertFalse(RequestValidation.validateOptionJournalModel.test(model));
//    }
//
//    @Test
//    void testValidateOptionJournalModelWhenStockPriceIsNegative() {
//        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
//        TransactionModel transactionModel = new TransactionModel("1234", "", OffsetDateTime.now(), "XYZ", TransactionType.OPTION, 6.1f, transactionSettingsModel);
//        OptionJournalModel model = new OptionJournalModel(transactionModel, -1, 2.1f, OffsetDateTime.now(), 1, 5.1f, ActionType.BUY, OptionType.PUT);
//
//        Assertions.assertFalse(RequestValidation.validateOptionJournalModel.test(model));
//    }
//
//    @Test
//    void testValidateOptionJournalModelWhenContractsIsZero() {
//        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
//        TransactionModel transactionModel = new TransactionModel("1234", "", OffsetDateTime.now(), "XYZ", TransactionType.OPTION, 6.1f, transactionSettingsModel);
//        OptionJournalModel model = new OptionJournalModel(transactionModel, 44.55, 2.1f, OffsetDateTime.now(), 0, 5.1f, ActionType.BUY, OptionType.PUT);
//
//        Assertions.assertFalse(RequestValidation.validateOptionJournalModel.test(model));
//    }
//
//    @Test
//    void testValidateOptionJournalModelWhenSymbolIsEmpty() {
//        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
//        TransactionModel transactionModel = new TransactionModel("1234", "", OffsetDateTime.now(), "", TransactionType.OPTION, 6.1f, transactionSettingsModel);
//        OptionJournalModel model = new OptionJournalModel(transactionModel, 44.55, 2.1f, OffsetDateTime.now(), 1, 5.1f, ActionType.BUY, OptionType.PUT);
//
//        Assertions.assertFalse(RequestValidation.validateOptionJournalModel.test(model));
//    }
//
//    @Test
//    void testValidateOptionJournalModelWhenSymbolIsSpace() {
//        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
//        TransactionModel transactionModel = new TransactionModel("1234", "", OffsetDateTime.now(), " ", TransactionType.OPTION, 6.1f, transactionSettingsModel);
//        OptionJournalModel model = new OptionJournalModel(transactionModel, 44.55, 2.1f, OffsetDateTime.now(), 1, 5.1f, ActionType.BUY, OptionType.PUT);
//
//        Assertions.assertFalse(RequestValidation.validateOptionJournalModel.test(model));
//    }
//
//    @Test
//    void testValidateOptionJournalModelTruthy() {
//        TransactionSettingsModel transactionSettingsModel = new TransactionSettingsModel("1234", 0.00, false, false);
//        TransactionModel transactionModel = new TransactionModel("1234", "123456789012345678901234567890123456", OffsetDateTime.now(), "XYZ", TransactionType.OPTION, 0, transactionSettingsModel);
//        OptionJournalModel model = new OptionJournalModel(transactionModel, 44.55, 55, OffsetDateTime.now(), 1, 1.5, ActionType.BUY, OptionType.CALL);
//
//        Assertions.assertTrue(RequestValidation.validateOptionJournalModel.test(model));
//    }
//}

File: services/tradelog/tradelog-api/src/test/kotlin/com/example/tradelog/api/rest/validator/ValidateOptionJournalModelTest.kt
package com.example.tradelog.api.rest.validator

import com.example.tradelog.api.spec.model.TLOptionJournalDto
import com.example.tradelog.api.spec.model.TLTransactionDto
import com.example.tradelog.api.spec.model.TLTransactionSettingsDto
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.time.OffsetDateTime

class ValidateOptionJournalModelTest {

    @Test
    fun testValidateOptionJournalModelWhenActionIsInvalid() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.OPTION)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val optionJournalDto = TLOptionJournalDto.newBuilder()
                .setTransactionDetails(transactionDto)
                .setOptionType(TLOptionJournalDto.OptionType.CALL)
                .setAction(TLOptionJournalDto.ActionType.UNKNOWN_ACTION_TYPE)
                .setContracts(1)
                .setExpiryDate(OffsetDateTime.now().toString())
                .setPremium(1.0)
                .setStockPrice(1.0)
                .setStrikePrice(1.0)
                .build()

        Assertions.assertFalse(RequestValidator.validateOptionJournalModel(optionJournalDto))
    }


    @Test
    fun testValidateOptionJournalModelWhenOptionTypeIsUnknown() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.OPTION)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val optionJournalDto = TLOptionJournalDto.newBuilder()
                .setTransactionDetails(transactionDto)
                .setOptionType(TLOptionJournalDto.OptionType.UNKNOWN_OPTION_TYPE)
                .setAction(TLOptionJournalDto.ActionType.SELL)
                .setContracts(1)
                .setExpiryDate(OffsetDateTime.now().toString())
                .setPremium(1.0)
                .setStockPrice(1.0)
                .setStrikePrice(1.0)
                .build()

        Assertions.assertFalse(RequestValidator.validateOptionJournalModel(optionJournalDto))
    }


    @Test
    fun testValidateOptionJournalModelWhenStockPriceIsNegative() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.OPTION)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val optionJournalDto = TLOptionJournalDto.newBuilder()
                .setTransactionDetails(transactionDto)
                .setOptionType(TLOptionJournalDto.OptionType.CALL)
                .setAction(TLOptionJournalDto.ActionType.SELL)
                .setContracts(1)
                .setExpiryDate(OffsetDateTime.now().toString())
                .setPremium(1.0)
                .setStockPrice(-1.0)
                .setStrikePrice(1.0)
                .build()

        Assertions.assertFalse(RequestValidator.validateOptionJournalModel(optionJournalDto))
    }


    @Test
    fun testValidateOptionJournalModelWhenContractsIsZero() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.OPTION)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val optionJournalDto = TLOptionJournalDto.newBuilder()
                .setTransactionDetails(transactionDto)
                .setOptionType(TLOptionJournalDto.OptionType.CALL)
                .setAction(TLOptionJournalDto.ActionType.SELL)
                .setContracts(0)
                .setExpiryDate(OffsetDateTime.now().toString())
                .setPremium(1.0)
                .setStockPrice(1.0)
                .setStrikePrice(1.0)
                .build()

        Assertions.assertFalse(RequestValidator.validateOptionJournalModel(optionJournalDto))
    }


    @Test
    fun testValidateOptionJournalModelWhenSymbolIsEmpty() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("")
                .setType(TLTransactionDto.TransactionType.OPTION)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val optionJournalDto = TLOptionJournalDto.newBuilder()
                .setTransactionDetails(transactionDto)
                .setOptionType(TLOptionJournalDto.OptionType.CALL)
                .setAction(TLOptionJournalDto.ActionType.SELL)
                .setContracts(1)
                .setExpiryDate(OffsetDateTime.now().toString())
                .setPremium(1.0)
                .setStockPrice(1.0)
                .setStrikePrice(1.0)
                .build()

        Assertions.assertFalse(RequestValidator.validateOptionJournalModel(optionJournalDto))
    }


    @Test
    fun testValidateOptionJournalModelWhenSymbolIsSpace() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol(" ")
                .setType(TLTransactionDto.TransactionType.OPTION)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val optionJournalDto = TLOptionJournalDto.newBuilder()
                .setTransactionDetails(transactionDto)
                .setOptionType(TLOptionJournalDto.OptionType.CALL)
                .setAction(TLOptionJournalDto.ActionType.SELL)
                .setContracts(1)
                .setExpiryDate(OffsetDateTime.now().toString())
                .setPremium(1.0)
                .setStockPrice(1.0)
                .setStrikePrice(1.0)
                .build()

        Assertions.assertFalse(RequestValidator.validateOptionJournalModel(optionJournalDto))
    }


    @Test
    fun testValidateOptionJournalModelWhenStrikePriceIsNegative() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.OPTION)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val optionJournalDto = TLOptionJournalDto.newBuilder()
                .setTransactionDetails(transactionDto)
                .setOptionType(TLOptionJournalDto.OptionType.CALL)
                .setAction(TLOptionJournalDto.ActionType.SELL)
                .setContracts(1)
                .setExpiryDate(OffsetDateTime.now().toString())
                .setPremium(1.0)
                .setStockPrice(1.0)
                .setStrikePrice(-1.0)
                .build()

        Assertions.assertFalse(RequestValidator.validateOptionJournalModel(optionJournalDto))
    }


    @Test
    fun testValidateOptionJournalModelTruthy() {
        val transactionSettingsDto = TLTransactionSettingsDto.newBuilder()
                .setTransactionId("1234")
                .setPreferredPrice(0.00)
                .setGroupSelected(false)
                .setLegClosed(false)
                .build()

        val transactionDto = TLTransactionDto.newBuilder()
                .setId("1234")
                .setDate(OffsetDateTime.now().toString())
                .setAccountId("123456789012345678901234567890123456")
                .setSymbol("XYZ")
                .setType(TLTransactionDto.TransactionType.OPTION)
                .setBrokerFees(0.00)
                .setSettings(transactionSettingsDto)
                .build()

        val optionJournalDto = TLOptionJournalDto.newBuilder()
                .setTransactionDetails(transactionDto)
                .setOptionType(TLOptionJournalDto.OptionType.CALL)
                .setAction(TLOptionJournalDto.ActionType.SELL)
                .setContracts(1)
                .setExpiryDate(OffsetDateTime.now().toString())
                .setPremium(1.0)
                .setStockPrice(1.0)
                .setStrikePrice(1.0)
                .build()

        Assertions.assertTrue(RequestValidator.validateOptionJournalModel(optionJournalDto))
    }
}